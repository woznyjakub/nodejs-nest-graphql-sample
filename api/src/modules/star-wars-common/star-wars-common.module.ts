import { Module } from '@nestjs/common';

import { StarWarsCommonService } from './star-wars-common.service';

@Module({
  providers: [StarWarsCommonService],
  exports: [StarWarsCommonService],
})
export class StarWarsCommonModule {}
