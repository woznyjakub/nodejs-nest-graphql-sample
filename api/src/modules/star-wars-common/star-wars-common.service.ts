import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';

export type PagedResponse<Item> = {
  count: number;
  next: string | null;
  previous: string | null;
  results: Item[];
};

@Injectable()
export class StarWarsCommonService {
  constructor(@Inject(CACHE_MANAGER) private cacheManager: Cache) {}

  async fetchWithCaching<T extends object>(
    cacheKey: string,
    getterFn: () => Promise<T>,
  ): Promise<T> {
    const cachedData = await this.cacheManager.get<T>(cacheKey);
    if (cachedData) {
      return cachedData;
    }

    const data = await getterFn();
    await this.cacheManager.set(cacheKey, data);

    return data;
  }

  async getAllPages<T extends object>(
    cacheHandlerFn: (page: number) => string,
    getterFn: (page: number) => Promise<PagedResponse<T>>,
  ): Promise<T[]> {
    const limit = 25;
    let page = 1;

    const results: T[] = [];

    do {
      const data = await this.fetchWithCaching(cacheHandlerFn(page), () => getterFn(page));

      if (Array.isArray(data.results) || !results?.length) {
        results.push(...data.results);
      } else {
        break;
      }

      if (!data.next) {
        break;
      }
      page++;
    } while (page < limit);

    return results;
  }

  private buildCacheKey(
    resourceName: string,
    dataType: 'page' | 'id',
    value: string | number,
  ): string {
    return `swapi-${resourceName}-${dataType}-${value}`;
  }

  getVehicleCacheKey(id: number): string {
    return this.buildCacheKey('vehicle', 'id', id);
  }

  getVehiclesCacheKey(page: number): string {
    return this.buildCacheKey('vehicles', 'page', page);
  }

  getFilmCacheKey(id: number): string {
    return this.buildCacheKey('film', 'id', id);
  }

  getFilmsCacheKey(page: number): string {
    return this.buildCacheKey('films', 'page', page);
  }

  getStarshipCacheKey(id: number): string {
    return this.buildCacheKey('starship', 'id', id);
  }

  getStarshipsCacheKey(page: number): string {
    return this.buildCacheKey('starships', 'page', page);
  }

  getPlanetCacheKey(id: number): string {
    return this.buildCacheKey('planet', 'id', id);
  }

  getPlanetsCacheKey(page: number): string {
    return this.buildCacheKey('planets', 'page', page);
  }

  getSpeciesCacheKey(id: number): string {
    return this.buildCacheKey('species', 'id', id);
  }

  getSpeciesManyCacheKey(page: number): string {
    return this.buildCacheKey('speciesMany', 'page', page);
  }

  getPeopleCacheKey(page: number): string {
    return this.buildCacheKey('people', 'page', page);
  }
}
