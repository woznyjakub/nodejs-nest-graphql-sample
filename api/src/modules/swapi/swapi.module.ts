import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';

import { SwapiService } from './swapi.service';

@Module({
  imports: [HttpModule],
  providers: [SwapiService],
  exports: [SwapiService],
})
export class SwapiModule {}
