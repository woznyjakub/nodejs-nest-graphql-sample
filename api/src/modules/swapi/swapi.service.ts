import { HttpService } from '@nestjs/axios';
import { HttpException, Injectable } from '@nestjs/common';
import { ClassConstructor, plainToInstance } from 'class-transformer';
import { validate } from 'class-validator';
import { lastValueFrom } from 'rxjs';

import { FilmResponse, FilmsResponse } from './dto/film';
import { PeopleResponse } from './dto/people';
import { PlanetResponse, PlanetsResponse } from './dto/planet';
import { SpeciesManyResponse, SpeciesResponse } from './dto/species';
import { StarshipResponse, StarshipsResponse } from './dto/starship';
import { VehicleResponse, VehiclesResponse } from './dto/vehicle';

@Injectable()
export class SwapiService {
  constructor(private readonly httpService: HttpService) {}

  private readonly baseUrl = 'https://swapi.dev/api';

  private async getResource<T extends object>(
    name: string,
    dto: ClassConstructor<T>,
    id?: number,
    page?: number,
  ): Promise<T> {
    const url = `${this.baseUrl}/${name}${id ? `/${id}` : ''}`;
    const params = !id && page ? { page } : null;

    try {
      const response = await lastValueFrom(this.httpService.get(url, { params }));
      const resourceData: unknown = response.data;

      const resource = plainToInstance<T, unknown>(dto, resourceData, {
        enableImplicitConversion: true,
        excludeExtraneousValues: true,
      });

      const errors = await validate(resource);
      if (errors.length > 0) {
        throw new Error(
          `Swapi resource validation failed: ${errors.map((e) => e.toString()).join(', ')}`,
        );
      }

      return resource;
    } catch (error) {
      throw new HttpException(
        `Failed to get swapi resource (${name}). Error: ${error.message}`,
        error.status,
      );
    }
  }

  getVehicle(id: number): Promise<VehicleResponse> {
    return this.getResource('vehicles', VehicleResponse, id);
  }

  getVehicles(page: number = 1): Promise<VehiclesResponse> {
    return this.getResource('vehicles', VehiclesResponse, undefined, page);
  }

  getFilm(id: number): Promise<FilmResponse> {
    return this.getResource('films', FilmResponse, id);
  }

  getFilms(page: number = 1): Promise<FilmsResponse> {
    return this.getResource('films', FilmsResponse, undefined, page);
  }

  getStarship(id: number): Promise<StarshipResponse> {
    return this.getResource('starships', StarshipResponse, id);
  }

  getStarships(page: number = 1): Promise<StarshipsResponse> {
    return this.getResource('starships', StarshipsResponse, undefined, page);
  }

  getPlanet(id: number): Promise<PlanetResponse> {
    return this.getResource('planets', PlanetResponse, id);
  }

  getPlanets(page: number = 1): Promise<PlanetsResponse> {
    return this.getResource('planets', PlanetsResponse, undefined, page);
  }

  getSpecies(id: number): Promise<SpeciesResponse> {
    return this.getResource('species', SpeciesResponse, id);
  }

  getSpeciesMany(page: number = 1): Promise<SpeciesManyResponse> {
    return this.getResource('species', SpeciesManyResponse, undefined, page);
  }

  getPeople(page: number = 1): Promise<PeopleResponse> {
    return this.getResource('people', PeopleResponse, undefined, page);
  }
}
