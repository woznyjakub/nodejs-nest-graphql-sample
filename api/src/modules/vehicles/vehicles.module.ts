import { Module } from '@nestjs/common';

import { StarWarsCommonModule } from '../star-wars-common/star-wars-common.module';
import { SwapiModule } from '../swapi/swapi.module';

import { VehiclesResolver } from './vehicles.resolver';
import { VehiclesService } from './vehicles.service';

@Module({
  imports: [SwapiModule, StarWarsCommonModule],
  providers: [VehiclesResolver, VehiclesService],
})
export class VehiclesModule {}
