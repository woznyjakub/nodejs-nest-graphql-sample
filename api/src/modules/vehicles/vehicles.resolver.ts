import { Resolver, Query, Args, Int } from '@nestjs/graphql';

import { Vehicle } from './models/vehicle.model';
import { Vehicles } from './models/vehicles.model';
import { VehiclesService } from './vehicles.service';

@Resolver(() => Vehicle)
export class VehiclesResolver {
  constructor(private readonly vehiclesService: VehiclesService) {}

  @Query(() => Vehicles, { name: 'vehicles' })
  findAll(@Args('page', { type: () => Int }) page: number): Promise<Vehicles> {
    return this.vehiclesService.findAll(page);
  }

  @Query(() => Vehicle, { name: 'vehicle' })
  findOne(@Args('id', { type: () => Int }) id: number): Promise<Vehicle> {
    return this.vehiclesService.findOne(id);
  }
}
