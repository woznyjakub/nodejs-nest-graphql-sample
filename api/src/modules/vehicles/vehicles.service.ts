import { Injectable } from '@nestjs/common';

import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { VehicleResponse } from '../swapi/dto/vehicle';
import { SwapiService } from '../swapi/swapi.service';

import { Vehicle } from './models/vehicle.model';
import { Vehicles } from './models/vehicles.model';

@Injectable()
export class VehiclesService {
  constructor(
    private readonly swapiService: SwapiService,
    private readonly swCommonService: StarWarsCommonService,
  ) {}

  async findAll(page: number): Promise<Vehicles> {
    const cacheKey = `swapi-vehicles-page-${page}`;

    const swapiData = await this.swCommonService.fetchWithCaching(cacheKey, () =>
      this.swapiService.getVehicles(page),
    );

    return {
      count: swapiData.count,
      next: swapiData.next,
      previous: swapiData.previous,
      results: swapiData.results.map(this.mapSingleVehicle),
    };
  }

  async findOne(id: number): Promise<Vehicle> {
    const cacheKey = `swapi-vehicle-id-${id}`;

    const swapiData = await this.swCommonService.fetchWithCaching(cacheKey, () =>
      this.swapiService.getVehicle(id),
    );

    return this.mapSingleVehicle(swapiData);
  }

  private mapSingleVehicle(vehicle: VehicleResponse): Vehicle {
    return {
      name: vehicle.name,
      model: vehicle.model,
      manufacturer: vehicle.manufacturer,
      costInCredits: vehicle.cost_in_credits,
      length: vehicle.length,
      maxAtmospheringSpeed: vehicle.max_atmosphering_speed,
      crew: vehicle.crew,
      passengers: vehicle.passengers,
      cargoCapacity: vehicle.cargo_capacity,
      consumables: vehicle.consumables,
      vehicleClass: vehicle.vehicle_class,
      pilots: vehicle.pilots,
      films: vehicle.films,
      url: vehicle.url,
    };
  }
}
