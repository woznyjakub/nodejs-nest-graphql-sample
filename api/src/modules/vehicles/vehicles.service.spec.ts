import { Test, TestingModule } from '@nestjs/testing';

import { vehicleResponseMock, vehiclesResponseMock } from '../../test/mocks/swapi';
import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { SwapiService } from '../swapi/swapi.service';

import { VehiclesService } from './vehicles.service';

describe('VehiclesService', () => {
  let vehiclesService: VehiclesService;

  const mockFetchWithCaching = vi.fn();

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        {
          provide: StarWarsCommonService,
          useValue: {
            fetchWithCaching: mockFetchWithCaching,
          },
        },
        VehiclesService,
        {
          provide: SwapiService,
          useValue: {
            getVehicle: vi.fn(),
            getVehicles: vi.fn(),
          },
        },
      ],
    }).compile();

    vehiclesService = module.get<VehiclesService>(VehiclesService);
  });

  describe('single vehicle data', () => {
    it('should return mapped data', async () => {
      mockFetchWithCaching.mockReturnValue(vehicleResponseMock);

      const result = await vehiclesService.findOne(1);

      expect(result).toMatchSnapshot();
    });
  });

  describe('multiple vehicles data', () => {
    it('should return mapped data', async () => {
      mockFetchWithCaching.mockReturnValue(vehiclesResponseMock);

      const result = await vehiclesService.findAll(1);

      expect(result).toMatchSnapshot();
    });
  });
});
