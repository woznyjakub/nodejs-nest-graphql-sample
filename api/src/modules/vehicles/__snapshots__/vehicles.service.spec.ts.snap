// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VehiclesService > multiple vehicles data > should return mapped data 1`] = `
{
  "count": 39,
  "next": "https://swapi.dev/api/vehicles/?page=2",
  "previous": null,
  "results": [
    {
      "cargoCapacity": "50000",
      "consumables": "2 months",
      "costInCredits": "150000",
      "crew": "46",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/5/",
      ],
      "length": "36.8 ",
      "manufacturer": "Corellia Mining Corporation",
      "maxAtmospheringSpeed": "30",
      "model": "Digger Crawler",
      "name": "Sand Crawler",
      "passengers": "30",
      "pilots": [],
      "url": "https://swapi.dev/api/vehicles/4/",
      "vehicleClass": "wheeled",
    },
    {
      "cargoCapacity": "50",
      "consumables": "0",
      "costInCredits": "14500",
      "crew": "1",
      "films": [
        "https://swapi.dev/api/films/1/",
      ],
      "length": "10.4 ",
      "manufacturer": "Incom Corporation",
      "maxAtmospheringSpeed": "1200",
      "model": "T-16 skyhopper",
      "name": "T-16 skyhopper",
      "passengers": "1",
      "pilots": [],
      "url": "https://swapi.dev/api/vehicles/6/",
      "vehicleClass": "repulsorcraft",
    },
    {
      "cargoCapacity": "5",
      "consumables": "unknown",
      "costInCredits": "10550",
      "crew": "1",
      "films": [
        "https://swapi.dev/api/films/1/",
      ],
      "length": "3.4 ",
      "manufacturer": "SoroSuub Corporation",
      "maxAtmospheringSpeed": "250",
      "model": "X-34 landspeeder",
      "name": "X-34 landspeeder",
      "passengers": "1",
      "pilots": [],
      "url": "https://swapi.dev/api/vehicles/7/",
      "vehicleClass": "repulsorcraft",
    },
    {
      "cargoCapacity": "65",
      "consumables": "2 days",
      "costInCredits": "unknown",
      "crew": "1",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "length": "6.4",
      "manufacturer": "Sienar Fleet Systems",
      "maxAtmospheringSpeed": "1200",
      "model": "Twin Ion Engine/Ln Starfighter",
      "name": "TIE/LN starfighter",
      "passengers": "0",
      "pilots": [],
      "url": "https://swapi.dev/api/vehicles/8/",
      "vehicleClass": "starfighter",
    },
    {
      "cargoCapacity": "10",
      "consumables": "none",
      "costInCredits": "unknown",
      "crew": "2",
      "films": [
        "https://swapi.dev/api/films/2/",
      ],
      "length": "4.5",
      "manufacturer": "Incom corporation",
      "maxAtmospheringSpeed": "650",
      "model": "t-47 airspeeder",
      "name": "Snowspeeder",
      "passengers": "0",
      "pilots": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/18/",
      ],
      "url": "https://swapi.dev/api/vehicles/14/",
      "vehicleClass": "airspeeder",
    },
    {
      "cargoCapacity": "none",
      "consumables": "2 days",
      "costInCredits": "unknown",
      "crew": "1",
      "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "length": "7.8",
      "manufacturer": "Sienar Fleet Systems",
      "maxAtmospheringSpeed": "850",
      "model": "TIE/sa bomber",
      "name": "TIE bomber",
      "passengers": "0",
      "pilots": [],
      "url": "https://swapi.dev/api/vehicles/16/",
      "vehicleClass": "space/planetary bomber",
    },
    {
      "cargoCapacity": "1000",
      "consumables": "unknown",
      "costInCredits": "unknown",
      "crew": "5",
      "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "length": "20",
      "manufacturer": "Kuat Drive Yards, Imperial Department of Military Research",
      "maxAtmospheringSpeed": "60",
      "model": "All Terrain Armored Transport",
      "name": "AT-AT",
      "passengers": "40",
      "pilots": [],
      "url": "https://swapi.dev/api/vehicles/18/",
      "vehicleClass": "assault walker",
    },
    {
      "cargoCapacity": "200",
      "consumables": "none",
      "costInCredits": "unknown",
      "crew": "2",
      "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "length": "2",
      "manufacturer": "Kuat Drive Yards, Imperial Department of Military Research",
      "maxAtmospheringSpeed": "90",
      "model": "All Terrain Scout Transport",
      "name": "AT-ST",
      "passengers": "0",
      "pilots": [
        "https://swapi.dev/api/people/13/",
      ],
      "url": "https://swapi.dev/api/vehicles/19/",
      "vehicleClass": "walker",
    },
    {
      "cargoCapacity": "10",
      "consumables": "1 day",
      "costInCredits": "75000",
      "crew": "2",
      "films": [
        "https://swapi.dev/api/films/2/",
      ],
      "length": "7",
      "manufacturer": "Bespin Motors",
      "maxAtmospheringSpeed": "1500",
      "model": "Storm IV Twin-Pod",
      "name": "Storm IV Twin-Pod cloud car",
      "passengers": "0",
      "pilots": [],
      "url": "https://swapi.dev/api/vehicles/20/",
      "vehicleClass": "repulsorcraft",
    },
    {
      "cargoCapacity": "2000000",
      "consumables": "Live food tanks",
      "costInCredits": "285000",
      "crew": "26",
      "films": [
        "https://swapi.dev/api/films/3/",
      ],
      "length": "30",
      "manufacturer": "Ubrikkian Industries Custom Vehicle Division",
      "maxAtmospheringSpeed": "100",
      "model": "Modified Luxury Sail Barge",
      "name": "Sail barge",
      "passengers": "500",
      "pilots": [],
      "url": "https://swapi.dev/api/vehicles/24/",
      "vehicleClass": "sail barge",
    },
  ],
}
`;

exports[`VehiclesService > single vehicle data > should return mapped data 1`] = `
{
  "cargoCapacity": "50000",
  "consumables": "2 months",
  "costInCredits": "150000",
  "crew": "46",
  "films": [
    "https://swapi.dev/api/films/1/",
  ],
  "length": "36.8",
  "manufacturer": "Corellia Mining Corporation",
  "maxAtmospheringSpeed": "30",
  "model": "Digger Crawler",
  "name": "Sand Crawler",
  "passengers": "30",
  "pilots": [],
  "url": "https://swapi.dev/api/vehicles/4/",
  "vehicleClass": "wheeled",
}
`;
